### ISSUE TITLE
Stuck when reloading domain

### ISSUE THREAD

**Issue by VincentMarnier** [Cysharp/YetAnotherHttpHandler#45]
Hello,
So this might partially be on my side, but from times to times (like 1 out of 6 times), my unity editor is stuck at "Reloading Domain".
If I look for the logs, I see:
```
Begin MonoManager ReloadAssembly
thread 'tokio-runtime-worker' panicked at yaha_native/src/binding.rs:84:85:
The context has already disposed: on_complete
stack backtrace:
Thread 0x32778f000 may have been prematurely finalized
Thread 0x326807000 may have been prematurely finalized
Thread 0x326e2b000 may have been prematurely finalized
   0:        0x325c55ff4 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h0524bd1b527203b8
   1:        0x325c71124 - core::fmt::write::h74be1333b0a6ddea
   2:        0x325c53ac8 - std::io::Write::write_fmt::h6c011b665b2d7d4b
   3:        0x325c55e34 - std::sys_common::backtrace::print::he25ff7dac6402434
   4:        0x325c57434 - std::panicking::default_hook::{{closure}}::h41763ba018b93bf7
   5:        0x325c5714c - std::panicking::default_hook::h017c44660373840e
   6:        0x325c57978 - std::panicking::rust_panic_with_hook::h221a0baf52ecd48a
   7:        0x325c57840 - std::panicking::begin_panic_handler::{{closure}}::h18b661cf0f97a414
   8:        0x325c56480 - std::sys_common::backtrace::__rust_end_short_backtrace::h4f4b2ffab2311217
   9:        0x325c575ec - _rust_begin_unwind
  10:        0x325c83a2c - core::panicking::panic_fmt::hd6070cd222575be5
  11:        0x325b2f47c - yaha_native::binding::_sentinel_on_complete::h9d141ccad330d382
  12:        0x325add314 - yaha_native::binding::yaha_request_begin::{{closure}}::h5631ce843d7c5f2a
  13:        0x325adb934 - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h3072105b67f01266
  14:        0x325b3cf2c - tokio::runtime::task::core::Core<T,S>::poll::hab7b4df7609e5b75
  15:        0x325ab8b5c - tokio::runtime::task::harness::Harness<T,S>::poll::hb04bc757ef901abc
  16:        0x325c191e0 - tokio::runtime::scheduler::multi_thread::worker::Context::run_task::habe660f9555d29f0
  17:        0x325c181a0 - tokio::runtime::scheduler::multi_thread::worker::Context::run::h92d568732cff005f
  18:        0x325c31ca8 - tokio::runtime::context::scoped::Scoped<T>::set::h28d62af9fe07fa31
  19:        0x325c31ba0 - tokio::runtime::context::runtime::enter_runtime::h21dc3e6654b89594
  20:        0x325c17d38 - tokio::runtime::scheduler::multi_thread::worker::run::h3240adbe89ca2def
  21:        0x325c1f72c - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h3f7f0a43009caa76
  22:        0x325c34a60 - tokio::runtime::task::core::Core<T,S>::poll::hac538a70219f2b02
  23:        0x325c14dd0 - tokio::runtime::task::harness::Harness<T,S>::poll::he5653619c802c7d6
  24:        0x325c1c41c - std::sys_common::backtrace::__rust_begin_short_backtrace::h2ef7fff63fe86b99
  25:        0x325c26374 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h2509ba3d429991b9
  26:        0x325c5976c - std::sys::unix::thread::Thread::new::thread_start::h07b00985b5bfc467
  27:        0x18a9ba034 - __pthread_joiner_wake
Thread 0x326207000 may have been prematurely finalized
```

---
How I create the channel:
```
GrpcChannelOptions options = new GrpcChannelOptions()
{
    Credentials = Config.ProjectsServiceCredentials,
    DisposeHttpClient = true,
    HttpHandler = new YetAnotherHttpHandler()
    {
        Http2Only = true
    }
};
_grpcChannel = GrpcChannel.ForAddress(Config.ProjectsServiceAddress, options);
```

---
How it is disposed:
```
_grpcChannel
  .ShutdownAsync()
  .ContinueWith(t =>
  {
      if (t.Exception != null)
      {
          UnityEngine.Debug.LogException(t.Exception);
      }
      _grpcChannel.Dispose();
  }, TaskScheduler.FromCurrentSynchronizationContext());
```

---
Steps to reproduce:
1. Click play button, wait until you are connected to the GRPC server
2. Click the stop button
3. Click the play button again

=> Expected behavior: it plays again
=> Actual behavior: editor is stuck at "Reloading Domain"

---
**EDIT:** If I replace my graceful dispose by just _grpcChannel.Dispose(), the issue still occurs
**EDIT2:** Same if I don't dispose it, except it is stuck 100% of times
**NOTE:** I am communicating with 3 GRPC servers and so use 3 YetAnotherHttpHandler

**Comment by mayuki** [comment]
Please tell us a few things so we can investigate in detail:

- Is the gRPC connection streaming? If streaming, is the streaming being disposed before the channel is disposed?
- When is a channel disposed? `OnApplicationQuit` event?
- Which Unity Editor version and platform?

**Comment by VincentMarnier** [comment]
Hello Mayuki,
I really like your input and I am pretty sure it has something to do with things not being disposed on time on my side.
Indeed I have streaming connections (from 3 to 8 depending on the use case). I just modified my code to be 100% sure they were disposed before disposing the channels but crashes still occurs.
I will keep modifying my code because if it works sometimes it must be possible to make it work every time and I am probably the one at fault here. I might missed a method called without CancellationToken or something just before disposing.
I will get back to you soon!

If that still matters:
The channel is disposed on OnApplicationQuit event.
I am using Unity 2022.3.18f1 on macOS Sonoma v14.2.1 with apple sillicon

**Comment by VincentMarnier** [comment]
I am coming back at you to keep you updated.

I spent quite some time trying to look for asynchronous issues I might be victim of and at some point I realized that I was instantiating 3 YetAnotherHttpHandler.
I thought "maybe deep down it tries to dispose of the same resource twice".
Now my 3 channels are sharing the same YetAnotherHttpHandler instance that I dispose once not needed anymore and I see no more crash.
Not sure if really fixed or if it just drastically reduced occurence. It did not happen in 1000 tests where it took about 4 previously.

I tried to reproduce the error in a minimal Unity Scene and failed to do so.
Might invest some more time in it later.
Thanks again for your input.
